File: ios/EarthUI/Sources/EarthUI/Govern/PieChart.swift
import SwiftUI

/// One option in an allocation stream, and the share given to it.
struct AllocationSlice: Identifiable, Equatable {
    let name: String
    let percent: Int

    var id: String { name }
}

/// A share of a whole, as a ring.
///
/// A ring rather than a filled pie: a reader comparing two of these compares
/// arc lengths either way, and the centre of a pie carries no information while
/// being the part hardest to judge by eye.
///
/// The hole used to state the allocated total. It was always "100%" for any
/// valid allocation — the slices are a split of the whole, so their sum is a
/// restatement of the thing the ring already draws — and it showed the
/// committed split rather than the one being dragged in the editor, so it read
/// as a figure that would not update.
///
/// Slices take `Palette.Series`, a fixed categorical order, because hue here is
/// doing identity work — which slice is which. They used to be one accent hue
/// stepped 16% in opacity per slice, which is a sequential ramp asked to encode
/// identity: adjacent slices differed by almost nothing and the fourth onward
/// were mud. The one-accent rule still governs the rest of the app; a chart's
/// interior is where hue has to mean something.
///
/// Arcs are separated by a hairline gap of surface. It reads as division
/// without a second colour, and it is also what lets the palette's tighter
/// pairs stand up under simulated colourblindness.
struct PieChart: View {
    @Environment(\.earth) private var theme
    let slices: [AllocationSlice]
    /// Written in the hole, and only worth it when there is nothing to draw: a
    /// stream nobody has voted in is an empty circle, which is otherwise
    /// indistinguishable from one that failed to load.
    var centreLabel: String?

    private static let size: CGFloat = 180
    private static let stroke: CGFloat = 34
    /// Surface showing between arcs. See the doc comment.
    private static let gapPoints: CGFloat = 2

    /// Sweeps between splits, so switching Actual to Preferred reads as the
    /// same chart changing rather than two charts.
    ///
    /// Starts at 1, not 0. It used to sweep in on first appear too, driven from
    /// `onAppear` — but a `withAnimation` issued in the frame a sheet is
    /// presenting gets dropped, which left `progress` at zero and the ring
    /// drawn with no sweep at all. An invisible chart is a worse trade than a
    /// missing flourish, and the animation's stated job — making the two lenses
    /// read as one chart — is untouched.
    @State private var progress: CGFloat = 1

    var body: some View {
        ZStack {
            Canvas { context, size in
                let total = max(slices.reduce(0) { $0 + $1.percent }, 1)
                let inset = Self.stroke / 2
                let box = CGRect(x: inset, y: inset,
                                 width: size.width - Self.stroke,
                                 height: size.height - Self.stroke)
                // Twelve o'clock, where a reader starts.
                var start = Angle.degrees(-90)
                let radius = box.width / 2
                // Two points of arc, as an angle. A single slice gets none —
                // there is nothing to divide it from, and a gap would read as a
                // ring that failed to close.
                let gap = slices.count > 1
                    ? Angle.degrees(Double(Self.gapPoints) / (2 * .pi * radius) * 360)
                    : Angle.degrees(0)

                for (index, slice) in slices.enumerated() {
                    let sweep = Angle.degrees(360 * Double(slice.percent) / Double(total) * progress)
                    // A slice thinner than the gap keeps its colour rather than
                    // vanishing: an option someone allocated to must still be
                    // findable next to its legend row.
                    let inner = sweep.degrees > gap.degrees ? sweep - gap : sweep
                    var arc = Path()
                    arc.addArc(
                        center: CGPoint(x: box.midX, y: box.midY),
                        radius: radius,
                        startAngle: start,
                        endAngle: start + inner,
                        clockwise: false
                    )
                    context.stroke(
                        arc,
                        with: .color(Palette.Series.slot(index)),
                        style: StrokeStyle(lineWidth: Self.stroke, lineCap: .butt)
                    )
                    start += sweep
                }
            }
            .frame(width: Self.size, height: Self.size)

            if let centreLabel {
                Text(centreLabel)
                    .font(EarthType.bodySmall)
                    .foregroundStyle(theme.colors.textTertiary)
            }
        }
        .frame(width: Self.size, height: Self.size)
        .onChange(of: slices) { _, _ in
            progress = 0
            withAnimation(.easeInOut(duration: 0.5)) { progress = 1 }
        }
    }
}

/// The chart's key.
///
/// Separate from the chart so a caller can put it beside or beneath depending
/// on width. Dots reuse the chart's stepped opacity, which is the only thing
/// tying a row to its arc.
struct PieLegend: View {
    @Environment(\.earth) private var theme
    let slices: [AllocationSlice]

    var body: some View {
        VStack(spacing: 0) {
            ForEach(Array(slices.enumerated()), id: \.element.id) { index, slice in
                HStack(spacing: theme.space.x4) {
                    Circle()
                        .fill(Palette.Series.slot(index))
                        .frame(width: 10, height: 10)
                    Text(slice.name)
                        .font(EarthType.bodySmall)
                        .foregroundStyle(theme.colors.textPrimary)
                        .lineLimit(2)
                        .frame(maxWidth: .infinity, alignment: .leading)
                    Text("\(slice.percent)%")
                        .font(EarthType.bodySmall).fontWeight(.semibold)
                        .foregroundStyle(theme.colors.textSecondary)
                }
                .padding(.vertical, theme.space.x4)
            }
            if slices.isEmpty {
                Spacer().frame(height: theme.space.x8)
                Text("Nothing allocated.")
                    .font(EarthType.bodySmall)
                    .foregroundStyle(theme.colors.textTertiary)
                    .frame(maxWidth: .infinity, alignment: .leading)
            }
        }
    }
}

File: android/app/src/main/java/com/example/earthwallet/ui/compose/PieChart.kt
package network.erth.wallet.ui.compose

import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.tween
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import network.erth.wallet.ui.theme.EarthAccent
import network.erth.wallet.ui.theme.EarthTheme
import network.erth.wallet.ui.vendor.theme.colors.EarthColors
import network.erth.wallet.ui.vendor.theme.typography.EarthTypography

/**
 * A share of a whole, as a ring.
 *
 * A ring rather than a filled pie: the hole gives the total somewhere to live,
 * and a reader comparing two of these compares arc lengths either way — the
 * centre of a pie carries no information and is the part hardest to judge by
 * eye.
 *
 * Slices are one hue at stepped opacity rather than a palette. The chart is
 * already labelled beside itself, so a second colour per slice would encode
 * nothing the legend does not, and the screen keeps one accent.
 */
@Composable
fun PieChart(
    slices: List<AllocationSlice>,
    modifier: Modifier = Modifier,
    /** Written in the hole. "100%" allocated, or "unallocated". */
    centreLabel: String? = null,
) {
    val total = slices.sumOf { it.percent }.coerceAtLeast(1)

    // Sweeps in rather than snapping, so switching between Actual and
    // Preferred reads as the same chart changing rather than two charts.
    val progress by animateFloatAsState(
        targetValue = 1f,
        animationSpec = tween(500),
        label = "pie",
    )

    Box(modifier.size(PIE_SIZE), contentAlignment = Alignment.Center) {
        Canvas(Modifier.size(PIE_SIZE)) {
            val stroke = PIE_STROKE.toPx()
            val inset = stroke / 2
            var start = -90f // twelve o'clock, where a reader starts
            slices.forEachIndexed { i, slice ->
                val sweep = 360f * slice.percent / total * progress
                drawArc(
                    color = EarthAccent.ink.stepped(i),
                    startAngle = start,
                    sweepAngle = sweep,
                    useCenter = false,
                    topLeft = androidx.compose.ui.geometry.Offset(inset, inset),
                    size = Size(size.width - stroke, size.height - stroke),
                    style = Stroke(width = stroke),
                )
                start += sweep
            }
        }
        if (centreLabel != null) {
            Text(
                text = centreLabel,
                style = EarthTypography.textSm,
                color = EarthColors.Text.textTertiary,
            )
        }
    }
}

/**
 * The chart's key.
 *
 * Separate from the chart so a caller can put it beside or beneath depending on
 * width. Dots reuse the chart's stepped opacity, which is the only thing tying
 * a row to its arc.
 */
@Composable
fun PieLegend(slices: List<AllocationSlice>, modifier: Modifier = Modifier) {
    val dimens = EarthTheme.dimens
    Column(modifier) {
        slices.forEachIndexed { i, slice ->
            Row(
                Modifier.fillMaxWidth().padding(vertical = dimens.space4),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Box(
                    Modifier
                        .size(10.dp)
                        .background(EarthAccent.ink.stepped(i), CircleShape),
                )
                Spacer(Modifier.padding(horizontal = dimens.space4))
                Text(
                    text = slice.name,
                    style = EarthTypography.textSm,
                    color = EarthColors.Text.textPrimary,
                    maxLines = 2,
                    overflow = TextOverflow.Ellipsis,
                    modifier = Modifier.weight(1f),
                )
                Text(
                    text = "${slice.percent}%",
                    style = EarthTypography.textSm,
                    fontWeight = FontWeight.SemiBold,
                    color = EarthColors.Text.textSecondary,
                )
            }
        }
        if (slices.isEmpty()) {
            Spacer(Modifier.height(dimens.space8))
            Text(
                text = "Nothing allocated.",
                style = EarthTypography.textSm,
                color = EarthColors.Text.textTertiary,
            )
        }
    }
}

private val PIE_SIZE = 180.dp
private val PIE_STROKE = 34.dp

/** Steps down the accent so adjacent slices separate without a second hue. */
internal fun Color.stepped(index: Int): Color =
    copy(alpha = (1f - index * 0.16f).coerceAtLeast(0.28f))
